base/deprecated.jl

```diff
@@ -56,41 +56,33 @@ macro deprecate(old,new)
     end
 end
 
+const have_dep_warned = Set()
+
 function depwarn(msg, funcsym)
     opts = JLOptions()
-    if opts.depwarn > 0
-        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+    if opts.depwarn == 1  # raise a warning
         fn = String(unsafe_load(cglobal(:jl_filename, Ptr{Cchar})))
+        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+        key = hash(string(fn, ln, msg))
+        (key in have_dep_warned) && return
```

Comment:
In general, using the filename and line number as keys is just wrong. They can be different for the same usage in the code and they can be the same for multiple usage calling from the same toplevel line.
The only solution I can think of is https://github.com/JuliaLang/julia/pull/16617#issuecomment-222231575